stko/optimizers/gulp.py

```diff
@@ -520,52 +557,131 @@ def _write_gulp_file(self, mol, metal_atoms, in_file, output_xyz):
+            cell_section = self._cell_section()
         else:
-            top_line = 'opti conv noautobond fix molmec cartesian\n'
+            # Constant volume.
+            top_line += 'conv '
+            cell_section = ''
+
+        top_line += 'noautobond fix molmec cartesian\n'
 
         position_section = self._position_section(mol, type_translator)
         bond_section = self._bond_section(mol, metal_atoms)
         species_section = self._species_section(type_translator)
 
         library = '\nlibrary uff4mof.lib\n'
 
+        if self._cell is not None:
+            output_cif = output_xyz.replace('xyz', 'cif')
+            periodic_output = f'output cif {output_cif}\n'
+        else:
+            periodic_output = ''
+
         output_section = (
             '\n'
+            f'maxcyc {self._maxcyc}\n'
             'terse inout potentials\n'
-            'terse inout cell\n'
+            'terse in cell\n'
             'terse in structure\n'
             'terse inout derivatives\n'
             f'output xyz {output_xyz}\n'
+            f'{periodic_output}'
             # 'output movie xyz steps_.xyz\n'
         )
 
         with open(in_file, 'w') as f:
             f.write(top_line)
+            f.write(cell_section)
             f.write(position_section)
             f.write(bond_section)
             f.write(species_section)
             f.write(library)
             f.write(output_section)
 
+    def update_cell(self, cif_filename):
+        """
+        Extract cell from optimised structure (in CIF).
+
+        Returns
+        -------
+        cell_info : :class:`dict`
+            Dictionary with cell definition.
+
+        """
+
+        cell_info = {}
+
+        targets = {
+            '_cell_length_a': 'a',
+            '_cell_length_b': 'b',
+            '_cell_length_c': 'c',
+            '_cell_angle_alpha': 'angle1',
+            '_cell_angle_beta': 'angle2',
+            '_cell_angle_gamma': 'angle3',
+        }
+
+        with open(cif_filename, 'r') as f:
+            lines = f.readlines()
+
+        for targ in targets:
+            for line in lines:
+                # Avoid running through the rest.
+                if targets[targ] in cell_info.keys():
+                    break
+                splits = line.rstrip().split(' ')
+                if splits[0] == targ:
+                    cell_info[targets[targ]] = float(splits[-1])
+
+        return cell_info
+
+    def _save_cif(self, filename, output_cif):
+        """
+        Save CIF from optimisation folder.
+
+        Parameters
+        ----------
+        filename : :class:`str`
+            The name of CIF file to be written.
+
+        Returns
+        -------
+        None : :class:`NoneType`
+
+        """
+
+        os.rename(f'{self._output_dir}/{output_cif}', filename)
+
     def assign_FF(self, mol):
         """
         Assign forcefield types to molecule.
 
         Parameters
         ----------
+        mol : :class:`.Molecule`
```

Comment:
Should this be `stk.Molecule` to explicitly define molecule types.

Reply:
Fixed